odlpp/cu_ntuples.py

```diff
@@ -0,0 +1,1458 @@
+
+        self._vector_impl = _TYPE_MAP_NPY2CUDA[self._dtype]
+
+    def element(self, inp=None, data_ptr=None):
+        """Create a new element.
+
+        Parameters
+        ----------
+        inp : `array-like` or scalar, optional
+            Input to initialize the new element.
+
+            If ``inp`` is a `numpy.ndarray` of shape ``(size,)``
+            and the same data type as this space, the array is wrapped,
+            not copied.
+            Other array-like objects are copied (with broadcasting
+            if necessary).
+
+            If a single value is given, it is copied to all entries.
+
+            If both ``inp`` and ``data_ptr`` are `None`, an empty
+            element is created with no guarantee of its state
+            (memory allocation only).
+
+        data_ptr : `int`, optional
+            Memory address of a CUDA array container
+
+            Cannot be combined with ``inp``.
+
+        Returns
+        -------
+        element : `CudaNtuplesVector`
+            The new element
+
+        Notes
+        -----
+        This method preserves "array views" of correct size and type,
+        see the examples below.
+
+        TODO: No, it does not yet!
+
+        Examples
+        --------
+        >>> uc3 = CudaNtuples(3, 'uint8')
+        >>> x = uc3.element(np.array([1, 2, 3], dtype='uint8'))
+        >>> x
+        CudaNtuples(3, 'uint8').element([1, 2, 3])
+        >>> y = uc3.element([1, 2, 3])
+        >>> y
+        CudaNtuples(3, 'uint8').element([1, 2, 3])
+        """
+        if inp is None:
+            if data_ptr is None:
+                return self.element_type(self, self._vector_impl(self.size))
+            else:  # TODO: handle non-1 length strides
+                return self.element_type(
+                    self, self._vector_impl.from_pointer(data_ptr, self.size,
+                                                         1))
+        else:
+            if data_ptr is None:
+                if isinstance(inp, self._vector_impl):
+                    return self.element_type(self, inp)
+                elif isinstance(inp, self.element_type):
+                    if inp.dtype == self.dtype:
+                        # Simply rewrap for matching dtypes
+                        return self.element_type(self, inp.data)
+                    else:
+                        # Bulk-copy for non-matching dtypes
+                        elem = self.element()
+                        elem[:] = inp
+                        return elem
+                else:
+                    # Array-like input. Need to go through a NumPy array
+                    arr = np.array(inp, copy=False, dtype=self.dtype, ndmin=1)
+                    if arr.shape != (self.size,):
+                        raise ValueError('expected input shape {}, got {}.'
+                                         ''.format((self.size,), arr.shape))
+                    elem = self.element()
+                    elem[:] = arr
```

Comment:
We still have different behavior for CPU and CUDA Ntuples: the CPU version [does not allow](https://github.com/odlgroup/odl/blob/94c4b28f676674bb7f4561d1c0832a49d79171d6/odl/space/numpy_ntuples.py#L121-L125) `inp` to be a single value (unless `size == 1`) while this code does. We should harmonize this. I'd say don't allow it.